src/ray/rpc/client_call.h

```diff
@@ -0,0 +1,165 @@
+#ifndef RAY_RPC_CLIENT_CALL_H
+#define RAY_RPC_CLIENT_CALL_H
+
+#include <grpcpp/grpcpp.h>
+#include <boost/asio.hpp>
+
+#include "ray/common/status.h"
+#include "ray/rpc/util.h"
+
+namespace ray {
+namespace rpc {
+
+/// Represents an outgoing gRPC request.
+///
+/// The lifecycle of a `ClientCall` is as follows.
+///
+/// When a client submits a new gRPC request, a new `ClientCall` object will be created
+/// by `ClientCallMangager::CreateCall`. Then the object will be used as the tag of
+/// `CompletionQueue`.
+///
+/// When the reply is received, `ClientCallMangager` will get the address of this object
+/// via `CompletionQueue`'s tag. And the manager should call `OnReplyReceived` and then
+/// delete this object.
+class ClientCall {
+ public:
+  /// The callback to be called by `ClientCallManager` when the reply of this request is
+  /// received.
+  virtual void OnReplyReceived() = 0;
+};
+
+class ClientCallManager;
+
+/// Reprents the client callback function of a particular rpc method.
+///
+/// \tparam Reply Type of the reply message.
+template <class Reply>
+using ClientCallback = std::function<void(const Status &status, const Reply &reply)>;
+
+/// Implementaion of the `ClientCall`. It represents a `ClientCall` for a particular
+/// RPC method.
+///
+/// \tparam Reply Type of the Reply message.
+template <class Reply>
+class ClientCallImpl : public ClientCall {
```

Comment:
Is there a reason to have both ClientCallImpl and ClientCall? It looks like they could be unified.

Reply:
The reason is because `ClientCallImpl` is templated. If we don't have a non-templated interface `ClientCall`, the user (`ClientCallManager`) has to be templated as well. And we'll need more boilerplate code.
The same reason also apply to `ServerCall`.
I'll add this explanation to the document.

Reply:
Ok yeah, that makes sense, thanks for the explanation!